global-script.lua

```diff
@@ -3245,4 +3410,282 @@ function tCompare(t1,t2)
+    if(a == b) then return end
+    swapPlayerAreaObjects(a, b)
+    swapPlayerAreaColors(a, b)
+    print(a .. " swapped places with " .. b .. ".")
+end
+
+function swapPlayerPresenceColors(fromColor, toColor)
+    if fromColor == toColor then return end 
+    local function initData(color, ix)
+        bag = getObjectFromGUID(PlayerBags[color])
+        return {
+            color = color,
+            ix = ix,
+            bag = bag,
+            qty = bag.getQuantity(),
+            tints = {},
+            objects = {},
+            pattern = color .. "'s (.*)",
+            bagContents = {},
+        }
+    end
+    local colors = {
+        from = initData(fromColor, 1),
+        to = initData(toColor, 2)
+    }
+
+    -- If both bags are intact, there's no real point in swapping them.
+    -- Just bail out fast.
+    if colors.from.qty == 20 and colors.to.qty == 20 then
+        return
+    end
+
+    readyTokens[fromColor], readyTokens[toColor] = readyTokens[toColor], readyTokens[fromColor]
+    selectedColors[fromColor], selectedColors[toColor] = selectedColors[toColor], selectedColors[fromColor]
+
+    -- Remove any items still in the bags
+    -- NOTE: TTS's documentation suggests we may need to wait a physics frame after doing this,
+    -- but this seems to work fine without doing that.  If something goes awry with this code in the future,
+    -- you might try re-adding the delay.
+    for color,data in pairs(colors) do
+        for i = 1,data.qty do 
+            local obj = data.bag.takeObject({
+                sound=false,
+                position={x=data.ix*2, z=200, y=i*2}    -- Chosen to be out-of-the-way and to prevent items from stacking.
+            })
+            table.insert(data.bagContents, obj)
+        end
+    end
+
+    -- Pass 1: Iterate over all objects looking for "<color>'s X".
+    -- Make a note of what we find and what tint it is.
+    local match = string.match  -- Performance
+    local name, suffix
+    for _,obj in pairs(getAllObjects()) do
+        name = obj.getName()
+        if name then
+            for _,data in pairs(colors) do 
+                suffix = match(name, data.pattern)
+                if suffix then
+                    data.tints[suffix] = obj.getColorTint()
+                    if not data.objects[suffix] then
+                        data.objects[suffix] = {obj}
+                    else
+                        table.insert(data.objects[suffix], obj)
+                    end
+                end
+            end
+        end
+    end
+
+    -- Pass 2: Iterate over found objects and swap color tints and object names.
+    -- After we're done, put objects in their new presence bag, if applicable.
+    for _,ab in pairs({{colors.from, colors.to}, {colors.to, colors.from}}) do
+        local a, b = unpack(ab)
+        for suffix, tint in pairs(a.tints) do
+            local newname = a.color .. "'s " .. suffix
+            for _, obj in ipairs(b.objects[suffix]) do
```

Comment:
this a bug around this line if you try to change color after grabbing a defend token from your bag

Reply:
This is now fixed.

Note that the colorswapped tokens cannot be returned to any defend bag.  (This is a new bug.)  Note that even in master, colorswapped tokens cannot be returned to their defend bag if their state has been changed (even if it is changed back).

There's some commented out code in the bugfix commit that actually deletes the defence tokens and respawns them in the same position/rotation from the new bag.  This works around the first issue, but I suspect a more permanent fix for the second issue would fix the first anyways.  (Realistically, dumping the defence tokens in ANY of the defend bags should just get rid of them.)

Reply:
Yea I've made a note to fix that myself. the element bags for example have code to delete any element dropped on it, I'll fix that for defend bags